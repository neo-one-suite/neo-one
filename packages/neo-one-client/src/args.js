/* @flow */
import BigNumber from 'bignumber.js';

import _ from 'lodash';
import { common } from '@neo-one/client-core';

import type {
  ABI,
  ABIEvent,
  ABIFunction,
  ABIParameter,
  ABIReturn,
  AddressString,
  AttributeArg,
  BlockFilter,
  BufferString,
  GetOptions,
  Hash160String,
  Hash256String,
  PublicKeyString,
  TransactionOptions,
} from './types';
import { InvalidArgumentError, InvalidNamedArgumentError } from './errors';

import { addressToScriptHash } from './helpers';
import converters from './user/converters';

export const assertString = (name: string, param: mixed): string => {
  if (param == null || typeof param !== 'string') {
    throw new InvalidArgumentError(
      `Expected string for ${name}, found: ${String(param)}`,
    );
  }

  return param;
};

export const assertNullableString = (name: string, param?: mixed): ?string => {
  if (param == null) {
    return param;
  }

  return assertString(name, param);
};

export const assertAddress = (address: mixed): AddressString => {
  if (address == null || typeof address !== 'string') {
    throw new InvalidArgumentError(
      `Address argument was not a string: ${String(address)}`,
    );
  }

  try {
    addressToScriptHash(address);
  } catch (error) {
    throw new InvalidArgumentError(`Invalid address: ${address}`);
  }

  return address;
};

export const assertHash160 = (hash: mixed): Hash160String => {
  const value = assertString('Hash160', hash);

  if (!value.startsWith('0x')) {
    throw new InvalidArgumentError(
      `Hash160 must start with '0x', found: ${String(hash)}`,
    );
  }

  try {
    common.stringToUInt160(value);
  } catch (error) {
    throw new InvalidArgumentError(
      `Invalid Hash160 param, found: ${String(hash)}`,
    );
  }

  return value;
};

export const assertHash256 = (hash: mixed): Hash256String => {
  const value = assertString('Hash256', hash);

  if (!value.startsWith('0x')) {
    throw new InvalidArgumentError(
      `Hash256 must start with '0x', found: ${String(hash)}`,
    );
  }

  try {
    common.stringToUInt256(value);
  } catch (error) {
    throw new InvalidArgumentError(
      `Invalid Hash256 param, found: ${String(hash)}`,
    );
  }

  return value;
};

export const assertBuffer = (buffer: mixed): BufferString => {
  const value = assertString('Buffer', buffer);
  if (Buffer.from(value, 'hex').toString('hex') !== value.toLowerCase()) {
    throw new InvalidArgumentError(
      `Expected hex string, found: ${String(buffer)}`,
    );
  }

  return value;
};

export const assertPublicKey = (publicKey: mixed): PublicKeyString => {
  const value = assertBuffer(publicKey);
  try {
    common.stringToECPoint(value);
    return value;
  } catch (error) {
    throw new InvalidArgumentError(
      `Expected PublicKey, found: ${String(publicKey)}`,
    );
  }
};

export const assertBigNumber = (value: mixed): BigNumber => {
  if (value == null || !BigNumber.isBigNumber(value)) {
    throw new InvalidArgumentError(
      `Expected BigNumber, found: ${String(value)}`,
    );
  }

  return (value: $FlowFixMe);
};

export const assertNullableBigNumber = (value: mixed): ?BigNumber => {
  if (value == null) {
    return value;
  }

  return assertBigNumber(value);
};

export const assertBoolean = (value: mixed): boolean => {
  if (value == null || typeof value !== 'boolean') {
    throw new InvalidArgumentError(`Expected boolean, found: ${String(value)}`);
  }

  return value;
};

export const assertNumber = (value: mixed): number => {
  if (value == null || typeof value !== 'number') {
    throw new InvalidArgumentError(`Expected number, found: ${String(value)}`);
  }

  return value;
};

export const assertNullableNumber = (value: mixed): ?number => {
  if (value != null && typeof value !== 'number') {
    throw new InvalidArgumentError(`Expected number, found: ${String(value)}`);
  }

  return value;
};

export const assertArray = (value: mixed): Array<mixed> => {
  if (!Array.isArray(value)) {
    throw new InvalidArgumentError(`Expected Array, found: ${String(value)}`);
  }

  return value;
};

export const assertBlockFilter = (filter: mixed): ?BlockFilter => {
  if (filter == null) {
    return filter;
  }

  if (typeof filter !== 'object') {
    throw new InvalidArgumentError(
      `Invalid BlockFilter param, found: ${String(filter)}`,
    );
  }

  if (_.isEmpty(filter)) {
    return ({}: $FlowFixMe);
  }

  if (filter.indexStart != null && typeof filter.indexStart !== 'number') {
    throw new InvalidArgumentError(
      `Invalid BlockFilter param, found: ${String(filter)}`,
    );
  }

  if (filter.indexStop != null && typeof filter.indexStop !== 'number') {
    throw new InvalidArgumentError(
      `Invalid BlockFilter param, found: ${String(filter)}`,
    );
  }

  if (
    filter.indexStart != null &&
    filter.indexStop != null &&
    filter.indexStart > filter.indexStop
  ) {
    throw new InvalidArgumentError(
      `Invalid BlockFilter param, found: ${String(filter)}`,
    );
  }

  return (filter: $FlowFixMe);
};

export const assertGetOptions = (options: mixed): ?GetOptions => {
  if (options == null) {
    return options;
  }

  if (typeof options !== 'object') {
    throw new InvalidArgumentError(
      `Invalid GetOptions param, found: ${String(options)}`,
    );
  }

  if (_.isEmpty(options)) {
    return ({}: $FlowFixMe);
  }

  if (options.timeout != null && typeof options.timeout !== 'number') {
    throw new InvalidArgumentError(
      `Invalid GetOptions param, found: ${String(options)}`,
    );
  }

  return (options: $FlowFixMe);
};

const ABI_TYPES = new Set([
  'Signature',
  'Boolean',
  'Hash160',
  'Hash256',
  'ByteArray',
  'PublicKey',
  'String',
  'InteropInterface',
  'Void',
  'Integer',
  'Array',
]);

export const assertABIReturn = (value: mixed): ABIReturn => {
  if (value == null || typeof value !== 'object') {
    throw new InvalidArgumentError(
      `Invalid ABI return, found: ${String(value)}`,
    );
  }

  if (!ABI_TYPES.has(value.type)) {
    throw new InvalidArgumentError(
      `Invalid ABI return, found: ${String(value)}`,
    );
  }

  if (value.type === 'Array') {
    assertABIReturn(value.value);
  } else if (value.type === 'Integer') {
    assertNumber(value.decimals);
  }

  return (value: $FlowFixMe);
};

export const assertABIParameter = (value: mixed): ABIParameter => {
  if (value == null || typeof value !== 'object') {
    throw new InvalidArgumentError(
      `Invalid ABI parameter, found: ${String(value)}`,
    );
  }

  assertString('name', value.name);
  assertABIReturn(value);
  return (value: $FlowFixMe);
};

export const assertABIFunction = (value: mixed): ABIFunction => {
  if (value == null || typeof value !== 'object') {
    throw new InvalidArgumentError(
      `Invalid ABI function, found: ${String(value)}`,
    );
  }

  assertString('name', value.name);
  if (value.constant != null) {
    assertBoolean(value.constant);
  }

  if (value.parameters != null) {
    assertArray(value.parameters).map((parameter) =>
      assertABIParameter(parameter),
    );
  }

  assertABIReturn(value.returnType);

  return (value: $FlowFixMe);
};

export const assertABIEvent = (value: mixed): ABIEvent => {
  if (value == null || typeof value !== 'object') {
    throw new InvalidArgumentError(
      `Invalid ABI event, found: ${String(value)}`,
    );
  }

  assertString('name', value.name);
  assertArray(value.parameters).map((parameter) =>
    assertABIParameter(parameter),
  );

  return (value: $FlowFixMe);
};

export const assertABI = (abi: mixed): ABI => {
  if (abi == null || typeof abi !== 'object') {
    throw new InvalidArgumentError(`Invalid ABI param, found: ${String(abi)}`);
  }

  assertArray(abi.functions).map((value) => assertABIFunction(value));
  if (abi.events != null) {
    assertArray(abi.events).map((event) => assertABIEvent(event));
  }

  return (abi: $FlowFixMe);
};

export const assertAttributeArg = (attribute?: mixed): AttributeArg => {
  try {
    converters.attribute((attribute: $FlowFixMe));
    return (attribute: $FlowFixMe);
  } catch (error) {
    throw new InvalidNamedArgumentError('AttributeArg', attribute);
  }
};

export const assertTransactionOptions = (
  options?: mixed,
): ?TransactionOptions => {
  if (options == null) {
    return options;
  }

  if (typeof options !== 'object') {
    throw new InvalidNamedArgumentError('TransactionOptions', options);
  }

  const { from } = options;
  if (from != null) {
    if (typeof from !== 'object') {
      throw new InvalidNamedArgumentError('TransactionOptions', options);
    }
    assertString('TransactionOptions', from.network);
    assertString('TransactionOptions', from.address);
  }

  if (options.attributes != null) {
    if (!Array.isArray(options.attributes)) {
      throw new InvalidNamedArgumentError('TransactionOptions', options);
    }
    options.attributes.forEach((attribute) => {
      assertAttributeArg(attribute);
    });
  }

  assertNullableBigNumber(options.networkFee);

  return (options: $FlowFixMe);
};
