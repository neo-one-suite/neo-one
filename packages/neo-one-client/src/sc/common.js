/* @flow */
import { type Param as ScriptBuilderParam } from '@neo-one/client-core';

import _ from 'lodash';

import type {
  Action,
  ActionRaw,
  ABIEvent,
  ABIParameter,
  ABIReturn,
  ContractParameter,
  EventParameters,
  InvocationResult,
  Param,
  RawInvocationResult,
} from '../types';
import {
  InvalidEventError,
  InvalidArgumentError,
  InvocationCallError,
} from '../errors';

import parameterConverters, { converters } from './parameters';
import paramCheckers from './params';

export const convertParameter = ({
  type,
  parameter,
}: {|
  type: ABIParameter | ABIReturn,
  parameter: ContractParameter,
|}): ?Param => parameterConverters[type.type](parameter, (type: $FlowFixMe));

export const getParametersObject = ({
  abiParameters,
  parameters,
}: {|
  abiParameters: Array<ABIParameter>,
  parameters: Array<ContractParameter>,
|}): EventParameters => {
  if (abiParameters.length !== parameters.length) {
    throw new InvalidArgumentError(
      `Expected ABI parameters length (${abiParameters.length}) to equal ` +
        `parameters length (${parameters.length})`,
    );
  }

  return _
    .zip(abiParameters, parameters)
    .reduce((acc, [abiParameter, parameter]) => {
      acc[abiParameter.name] = convertParameter({
        type: abiParameter,
        parameter,
      });
      return acc;
    }, {});
};

export const convertAction = ({
  action,
  events,
}: {|
  action: ActionRaw,
  events: { [key: string]: ABIEvent },
|}): Action => {
  if (action.type === 'Log') {
    return action;
  }

  const { args } = action;
  if (args.length === 0) {
    throw new InvalidEventError('Notification had no arguments');
  }

  const event = converters.toString(args[0]);
  const eventSpec = events[event];
  if (eventSpec == null) {
    throw new InvalidEventError(`Unknown event ${event}`);
  }

  return {
    version: action.version,
    blockIndex: action.blockIndex,
    blockHash: action.blockHash,
    transactionIndex: action.transactionIndex,
    transactionHash: action.transactionHash,
    index: action.index,
    globalIndex: action.globalIndex,
    scriptHash: action.scriptHash,
    type: 'Event',
    name: event,
    parameters: getParametersObject({
      abiParameters: eventSpec.parameters,
      parameters: args.slice(1),
    }),
  };
};

export const convertInvocationResult = ({
  returnType,
  result,
}: {|
  returnType: ABIReturn,
  result: RawInvocationResult,
|}): InvocationResult<?Param> => {
  const { gasConsumed, gasCost } = result;
  if (result.state === 'FAULT') {
    return {
      state: result.state,
      gasConsumed,
      gasCost,
      message: result.message,
    };
  }

  let value = result.stack[0];
  if (value != null) {
    value = convertParameter({
      type: returnType,
      parameter: value,
    });
  }

  return { state: result.state, gasConsumed, gasCost, value };
};

export const convertCallResult = ({
  returnType,
  result,
}: {|
  returnType: ABIReturn,
  result: RawInvocationResult,
|}): ?Param => {
  if (result.state === 'FAULT') {
    throw new InvocationCallError(result.message);
  }

  let value = result.stack[0];
  if (value != null) {
    value = convertParameter({
      type: returnType,
      parameter: value,
    });
  }

  return value;
};

export const convertParams = ({
  parameters,
  params,
}: {|
  parameters: Array<ABIParameter>,
  params: Array<?Param>,
|}): {|
  converted: Array<?ScriptBuilderParam>,
  zipped: Array<[string, ?Param]>,
|} => {
  if (parameters.length !== params.length) {
    throw new InvalidArgumentError(
      `Expected parameters length (${parameters.length}) to equal params ` +
        `length (${params.length}).`,
    );
  }

  const converted = _
    .zip(parameters, params)
    .map(([parameter, param]) =>
      paramCheckers[parameter.type](param, (parameter: $FlowFixMe)),
    );
  const zipped = _
    .zip(parameters, params)
    .map(([parameter, param]) => [parameter.name, param]);

  return { converted, zipped };
};
