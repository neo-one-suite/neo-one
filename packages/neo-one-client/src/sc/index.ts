export * from './createReadSmartContract';
export * from './createSmartContract';
