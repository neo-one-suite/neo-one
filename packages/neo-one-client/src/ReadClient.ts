import { Param as ScriptBuilderParam } from '@neo-one/client-core';
import { Monitor } from '@neo-one/monitor';
import * as args from './args';
import { createReadSmartContract } from './sc';
import {
  Account,
  ActionRaw,
  AddressString,
  Asset,
  Block,
  BlockFilter,
  BufferString,
  Contract,
  DataProvider,
  GetOptions,
  Hash160String,
  Hash256String,
  Input,
  Output,
  Peer,
  RawCallReceipt,
  ReadSmartContract,
  ReadSmartContractDefinition,
  StorageItem,
  Transaction,
  Validator,
} from './types';

export class ReadClient<TDataProvider extends DataProvider = DataProvider> {
  public readonly dataProvider: TDataProvider;

  public constructor(dataProvider: TDataProvider) {
    this.dataProvider = dataProvider;
  }

  public async getAccount(address: AddressString, monitor?: Monitor): Promise<Account> {
    args.assertAddress(address);

    return this.dataProvider.getAccount(address, monitor);
  }

  public async getAsset(hash: Hash256String, monitor?: Monitor): Promise<Asset> {
    args.assertHash256(hash);

    return this.dataProvider.getAsset(hash, monitor);
  }

  public async getBlock(hash: number | Hash256String, options?: GetOptions): Promise<Block> {
    args.assertGetOptions(options);

    if (typeof hash === 'number') {
      return this.dataProvider.getBlock(hash, options);
    }

    return this.dataProvider.getBlock(args.assertHash256(hash));
  }

  public iterBlocks(filter?: BlockFilter): AsyncIterable<Block> {
    args.assertBlockFilter(filter);

    return this.dataProvider.iterBlocks(filter);
  }

  public async getBestBlockHash(monitor?: Monitor): Promise<Hash256String> {
    return this.dataProvider.getBestBlockHash(monitor);
  }

  public async getBlockCount(monitor?: Monitor): Promise<number> {
    return this.dataProvider.getBlockCount(monitor);
  }

  public async getContract(hash: Hash160String, monitor?: Monitor): Promise<Contract> {
    args.assertHash160(hash);

    return this.dataProvider.getContract(hash, monitor);
  }

  public async getMemPool(monitor?: Monitor): Promise<ReadonlyArray<Hash256String>> {
    return this.dataProvider.getMemPool(monitor);
  }

  public async getTransaction(hash: Hash256String, monitor?: Monitor): Promise<Transaction> {
    args.assertHash256(hash);

    return this.dataProvider.getTransaction(hash, monitor);
  }

  public async getOutput(input: Input, monitor?: Monitor): Promise<Output> {
    return this.dataProvider.getOutput(input, monitor);
  }

  public async getValidators(monitor?: Monitor): Promise<ReadonlyArray<Validator>> {
    return this.dataProvider.getValidators(monitor);
  }

  public async getConnectedPeers(monitor?: Monitor): Promise<ReadonlyArray<Peer>> {
    return this.dataProvider.getConnectedPeers(monitor);
  }

  public smartContract(definition: ReadSmartContractDefinition): ReadSmartContract {
    args.assertHash160(definition.hash);
    args.assertABI(definition.abi);

    return createReadSmartContract({ definition, client: this });
  }

  public async getStorage(hash: Hash160String, key: BufferString, monitor?: Monitor): Promise<StorageItem> {
    args.assertHash160(hash);
    args.assertBuffer(key);

    return this.dataProvider.getStorage(hash, key, monitor);
  }

  public iterStorage(hash: Hash160String, monitor?: Monitor): AsyncIterable<StorageItem> {
    args.assertHash160(hash);

    return this.dataProvider.iterStorage(hash, monitor);
  }

  public iterActionsRaw(filter?: BlockFilter): AsyncIterable<ActionRaw> {
    args.assertBlockFilter(filter);

    return this.dataProvider.iterActionsRaw(filter);
  }

  public async call(
    contract: Hash160String,
    method: string,
    params: ReadonlyArray<ScriptBuilderParam | undefined>,
    monitor?: Monitor,
  ): Promise<RawCallReceipt> {
    return this.dataProvider.call(contract, method, params, monitor);
  }
}
