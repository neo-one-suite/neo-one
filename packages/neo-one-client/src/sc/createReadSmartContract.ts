import { ScriptBuilderParam } from '@neo-one/client-core';
import { Monitor } from '@neo-one/monitor';
import { AsyncIterableX } from '@reactivex/ix-es2015-cjs/asynciterable/asynciterablex';
import { filter } from '@reactivex/ix-es2015-cjs/asynciterable/pipe/filter';
import { map } from '@reactivex/ix-es2015-cjs/asynciterable/pipe/map';
import BigNumber from 'bignumber.js';
import { ReadClient } from '../ReadClient';
import {
  ABIEvent,
  ABIFunction,
  ABIParameter,
  Action,
  AddressString,
  BlockFilter,
  Event,
  Log,
  Param,
  RawAction,
  ReadSmartContractAny,
  ReadSmartContractDefinition,
  SourceMaps,
  StorageItem,
} from '../types';
import * as common from './common';

const getParams = ({
  parameters,
  params: paramsIn,
}: {
  readonly parameters: ReadonlyArray<ABIParameter>;
  // tslint:disable-next-line no-any
  readonly params: ReadonlyArray<any>;
}): { readonly params: ReadonlyArray<ScriptBuilderParam | undefined>; readonly monitor?: Monitor } => {
  let params = paramsIn;
  const finalArg = params[params.length - 1];
  let monitor;
  if (
    finalArg != undefined &&
    typeof finalArg === 'object' &&
    !Array.isArray(finalArg) &&
    !BigNumber.isBigNumber(finalArg)
  ) {
    params = params.slice(0, -1);
    monitor = finalArg;
  }

  const { converted } = common.convertParams({ params, parameters });

  return { params: converted, monitor };
};

const createCall = ({
  address,
  client,
  func: { name, parameters = [], returnType },
  sourceMaps,
}: {
  readonly address: AddressString;
  readonly client: ReadClient;
  readonly func: ABIFunction;
  readonly sourceMaps?: SourceMaps;
  // tslint:disable-next-line no-any
}) => async (...args: any[]): Promise<Param | undefined> => {
  const { params, monitor } = getParams({
    parameters,
    params: args,
  });

  const receipt = await client.__call(address, name, params, monitor);

  return common.convertCallResult({ returnType, result: receipt.result, actions: receipt.actions, sourceMaps });
};

export const createReadSmartContract = ({
  definition,
  client,
}: {
  readonly definition: ReadSmartContractDefinition;
  readonly client: ReadClient;
}): ReadSmartContractAny => {
  const {
    address,
    abi: { events: abiEvents = [], functions },
    sourceMaps,
  } = definition;
  const events = abiEvents.reduce<{ [key: string]: ABIEvent }>(
    (acc, event) => ({
      ...acc,
      [event.name]: event,
    }),
    {},
  );

  const iterActionsRaw = (blockFilter: BlockFilter = {}): AsyncIterable<RawAction> =>
    AsyncIterableX.from(client.__iterActionsRaw(blockFilter)).pipe(filter((action) => action.address === address));

  const convertAction = (action: RawAction): Action => common.convertAction({ action, events });

  const iterActions = (filterIn?: BlockFilter): AsyncIterable<Action> =>
    AsyncIterableX.from(iterActionsRaw(filterIn)).pipe(map(convertAction));

  const iterEvents = (actionFilter?: BlockFilter): AsyncIterable<Event> =>
    AsyncIterableX.from(iterActions(actionFilter)).pipe(
      map((action) => {
        if (action.type === 'Log') {
          return undefined;
        }

        return action;
      }),
      filter(Boolean),
    );

  const iterLogs = (actionFilter?: BlockFilter): AsyncIterable<Log> =>
    AsyncIterableX.from(iterActions(actionFilter)).pipe(
      map((action) => {
        if (action.type === 'Event') {
          return undefined;
        }

        return action;
      }),
      filter(Boolean),
    );

  const iterStorage = (): AsyncIterable<StorageItem> => client.iterStorage(address);

  return functions.reduce<ReadSmartContractAny>(
    (acc, func) =>
      func.constant === true
        ? {
            ...acc,
            [func.name]: createCall({ client, address, func, sourceMaps }),
          }
        : acc,
    {
      definition,
      iterActions,
      iterEvents,
      iterLogs,
      iterStorage,
      convertAction,
    },
  );
};
