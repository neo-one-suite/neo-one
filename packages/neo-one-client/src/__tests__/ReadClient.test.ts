import * as assertArgs from '../args';
import { ReadClient } from '../ReadClient';
import { createReadSmartContract } from '../sc/createReadSmartContract';

jest.mock('../sc/createReadSmartContract');

describe('ReadClient', () => {
  let readClient = new ReadClient({} as any);
  let provider = readClient.dataProvider;
  beforeEach(() => {
    readClient = new ReadClient({} as any);
    provider = readClient.dataProvider;
  });

  const expected = '10';
  const dummyArg = 5;

  const testCases = [
    {
      method: 'getAccount',
      asserts: ['assertAddress'],
      args: [dummyArg],
    },

    {
      method: 'getAsset',
      asserts: ['assertHash256'],
      args: [dummyArg],
    },

    {
      method: 'getBlock',
      asserts: ['assertHash256'],
      args: [dummyArg],
    },

    {
      method: 'getBlock',
      asserts: ['assertGetOptions'],
      args: [dummyArg.toString()],
    },

    {
      method: 'iterBlocks',
      asserts: ['assertBlockFilter'],
      args: [dummyArg],
    },

    {
      method: 'getBestBlockHash',
      asserts: [],
      args: [dummyArg],
    },

    {
      method: 'getBlockCount',
      asserts: [],
      args: [dummyArg],
    },

    {
      method: 'getContract',
      asserts: ['assertHash160'],
      args: [dummyArg],
    },

    {
      method: 'getMemPool',
      asserts: [],
      args: [dummyArg],
    },

    {
      method: 'getTransaction',
      asserts: ['assertHash256'],
      args: [dummyArg],
    },

    {
      method: 'getValidators',
      asserts: [],
      args: [dummyArg],
    },

    {
      method: 'getConnectedPeers',
      asserts: [],
      args: [dummyArg],
    },

    {
      method: 'getStorage',
      asserts: ['assertHash160', 'assertBuffer'],
      args: [dummyArg],
    },

    {
      method: 'iterStorage',
      asserts: ['assertHash160'],
      args: [dummyArg],
    },

    {
      method: 'iterActionsRaw',
      asserts: ['assertBlockFilter'],
      args: [dummyArg],
    },

    {
      method: 'call',
      asserts: [],
      args: [dummyArg, dummyArg.toString(), [dummyArg]],
    },
  ];

  for (const testCase of testCases) {
    const { method, asserts, args } = testCase;
    let providerMethod = method;
    if (method.charAt(0) === '_') {
      providerMethod = method.slice(1);
    }

    test(method, async () => {
      for (const assert of asserts) {
        (assertArgs as any)[assert] = jest.fn(() => expected);
      }

      provider[providerMethod] = jest.fn(() => expected);

      const result = (readClient as any)[method](...args);

      expect(result).toEqual(expected);
    });
  }

  test('smartContract', () => {
    (assertArgs as any).assertHash160 = jest.fn(() => expected);
    (assertArgs as any).assertABI = jest.fn(() => expected);

    (createReadSmartContract as any).mockImplementation(() => expected);

    const result = readClient.smartContract(dummyArg.toString(), dummyArg as any);

    expect(result).toEqual(expected);
  });
});
