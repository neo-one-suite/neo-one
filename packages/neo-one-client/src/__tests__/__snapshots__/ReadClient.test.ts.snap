// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadClient __call 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient __iterActionsRaw 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getAccount 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    undefined,
  ],
]
`;

exports[`ReadClient getAsset 1`] = `
Array [
  Array [
    "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
    undefined,
  ],
]
`;

exports[`ReadClient getBestBlockHash 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getBlock - number 1`] = `
Array [
  Array [
    0,
    undefined,
  ],
]
`;

exports[`ReadClient getBlock 1`] = `
Array [
  Array [
    "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
    undefined,
  ],
]
`;

exports[`ReadClient getBlockCount 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getConnectedPeers 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getContract 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    undefined,
  ],
]
`;

exports[`ReadClient getMemPool 1`] = `
Array [
  Array [
    undefined,
  ],
]
`;

exports[`ReadClient getOutput 1`] = `
Array [
  Array [
    Object {
      "hash": "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
      "index": 0,
    },
    undefined,
  ],
]
`;

exports[`ReadClient getStorage 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    "b500",
    undefined,
  ],
]
`;

exports[`ReadClient getTransaction 1`] = `
Array [
  Array [
    "0xb50034d97ba8c836758de1124b6c77d38bc772abc9a8f22c85e7389014e75234",
    undefined,
  ],
]
`;

exports[`ReadClient iterBlocks 1`] = `
Array [
  Array [
    Object {
      "indexStart": 3,
      "indexStop": 4,
    },
  ],
]
`;

exports[`ReadClient iterStorage 1`] = `
Array [
  Array [
    "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    "b500",
    undefined,
  ],
]
`;

exports[`ReadClient smartContract 1`] = `
Object {
  "balanceOf": [Function],
  "convertAction": [Function],
  "decimals": [Function],
  "definition": Object {
    "abi": Object {
      "events": Array [
        Object {
          "name": "transfer",
          "parameters": Array [
            Object {
              "name": "from",
              "optional": undefined,
              "type": "Address",
            },
            Object {
              "name": "to",
              "optional": undefined,
              "type": "Address",
            },
            Object {
              "decimals": 8,
              "name": "amount",
              "optional": undefined,
              "type": "Integer",
            },
          ],
        },
      ],
      "functions": Array [
        Object {
          "constant": true,
          "name": "name",
          "parameters": Array [],
          "returnType": Object {
            "optional": undefined,
            "type": "String",
          },
          "verify": undefined,
        },
        Object {
          "constant": true,
          "name": "symbol",
          "parameters": Array [],
          "returnType": Object {
            "optional": undefined,
            "type": "String",
          },
          "verify": undefined,
        },
        Object {
          "constant": true,
          "name": "decimals",
          "parameters": Array [],
          "returnType": Object {
            "decimals": 0,
            "optional": undefined,
            "type": "Integer",
          },
          "verify": undefined,
        },
        Object {
          "constant": true,
          "name": "totalSupply",
          "parameters": Array [],
          "returnType": Object {
            "decimals": 8,
            "optional": undefined,
            "type": "Integer",
          },
          "verify": undefined,
        },
        Object {
          "constant": undefined,
          "name": "transfer",
          "parameters": Array [
            Object {
              "name": "from",
              "optional": undefined,
              "type": "Address",
            },
            Object {
              "name": "to",
              "optional": undefined,
              "type": "Address",
            },
            Object {
              "decimals": 8,
              "name": "value",
              "optional": undefined,
              "type": "Integer",
            },
          ],
          "returnType": Object {
            "optional": undefined,
            "type": "Boolean",
          },
          "verify": undefined,
        },
        Object {
          "constant": true,
          "name": "balanceOf",
          "parameters": Array [
            Object {
              "name": "account",
              "optional": undefined,
              "type": "Address",
            },
          ],
          "returnType": Object {
            "decimals": 8,
            "optional": undefined,
            "type": "Integer",
          },
          "verify": undefined,
        },
      ],
    },
    "address": "ALq7AWrhAueN6mJNqk6FHJjnsEoPRytLdW",
    "sourceMaps": undefined,
  },
  "iterActions": [Function],
  "iterEvents": [Function],
  "iterLogs": [Function],
  "iterStorage": [Function],
  "name": [Function],
  "symbol": [Function],
  "totalSupply": [Function],
}
`;
